File: QuantConnect/Strategy_1.cs


# TrendFollowingADX.cs

using System;
using System.Linq;
using QuantConnect.Data;
using QuantConnect.Indicators;

namespace QuantConnect.Algorithm.CSharp
{
    public class TrendFollowingADX : QCAlgorithm
    {
        private AverageDirectionalIndex _adx;
        private RollingWindow<IndicatorDataPoint> _adxWindow;

        private Security _security;

        public override void Initialize()
        {
            SetStartDate(2015, 1, 1);
            SetEndDate(DateTime.Now.Date.AddDays(-1));
            SetCash(100000);

            _security = AddEquity("SPY", Resolution.Daily).Symbol;

            _adx = ADX(_security, 14);

            _adxWindow = new RollingWindow<IndicatorDataPoint>(2);
        }

        public override void OnData(Slice data)
        {
            if (!data.ContainsKey(_security))
                return;

            var price = data[_security].Close;

            if (!_adx.IsReady)
                return;

            _adxWindow.Add(_adx.Current);

            if (_adxWindow.Count < 2)
                return;

            var previousAdx = _adxWindow[0];
            var currentAdx = _adxWindow[1];

            var above20 = currentAdx.Value > 20;
            var below20 = currentAdx.Value < 20;
            
			var isLongPosition = Portfolio[_security].IsLong;
			var isShortPosition = Portfolio[_security].IsShort;
			var isFlatPosition = !isLongPosition && !isShortPosition;

			if (above20 && isFlatPosition)
			{
				SetHoldings(_security, 1.0);
				Debug($"BUY {_security} at {price}");
			}
			else if (below20 && isFlatPosition)
			{
				SetHoldings(_security, -1.0);
				Debug($"SHORT {_security} at {price}");
			}
			else if ((below20 && isLongPosition) || (above20 && isShortPosition))
			{
				Liquidate(_security);
				Debug($"EXIT {_security} at {price}");
			}

        	if (isLongPosition || isShortPosition)
        	{
            	var highPrice = data[_security].High;
            	var lowPrice = data[_security].Low;

            	if (isLongPosition && currentAdx.Value < previousAdx.Value && highPrice < price - ATR(_security, 14).Current * 2)
            	{
                	Liquidate(_security);
                	Debug($"STOP LOSS LONG {_security} at {price}");
            	}
            	else if (isShortPosition && currentAdx.Value < previousAdx.Value && lowPrice > price + ATR(_security, 14).Current * 2)
            	{
                	Liquidate(_security);
                	Debug($"STOP LOSS SHORT {_security} at {price}");
            	}
        	}
    	}
    }
}

File: QuantConnect/Strategy_1.py
# TrendFollowing_ADX.py

from datetime import timedelta
from typing import Type

from QuantConnect import QCAlgorithm
from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import (
    AverageDirectionalIndex,
    AverageTrueRange,
    ExponentialMovingAverage,
    RelativeStrengthIndex,
)
from QuantConnect.Orders import OrderStatus, UpdateOrderFields


class TrendFollowingADX(QCAlgorithm):
    def Initialize(self):
        self.SetStartDate(2015, 1, 1)
        self.SetEndDate(2020, 1, 1)
        self.SetCash(100000)

        # Add equity
        equity = self.AddEquity("SPY", Resolution.Daily)

        # Set up indicators
        self.adx = AverageDirectionalIndex(14)
        self.ema20 = ExponentialMovingAverage(20)
        self.rsi = RelativeStrengthIndex(14)
        self.atr = AverageTrueRange(14)

    def OnData(self, data: TradeBar):
        if not (
            self.adx.IsReady and self.ema20.IsReady and self.rsi.IsReady and self.atr.IsReady
        ):
            return

        # Check for entry criteria
        if not self.Portfolio.Invested and (
            self.adx.Current.Value > 20
        ) and (
            data.Close > self.ema20.Current.Value
        ) and (
            self.rsi.Current.Value > 50
        ):
            stopPrice = data.Close - 2 * self.atr.Current.Value
            quantity = int(self.Portfolio.Cash / data.Close)
            if quantity == 0:
                return

            # Place buy order with stop loss
            orderTicket = self.MarketOnCloseOrder("SPY", quantity, stopPrice)

            if orderTicket.Status != OrderStatus.Filled:
                return

            # Set trailing stop loss
            updateFields = UpdateOrderFields()
            updateFields.StopPrice = stopPrice * 0.95
            updateFields.Tag = "TrailingStopLoss"
            updateFields.Quantity = quantity
            updateFields.Symbol = "SPY"
            updateFields.OrderType = OrderType.StopMarket

            orderTicket.Update(updateFields)

        # Check for exit criteria
        elif self.Portfolio.Invested:

            holdingPeriodDays = (self.Time - self.Portfolio.InvestedTime).days

            if holdingPeriodDays >= 5:
                stopLossOrderTickets = [
                    x
                    for x in self.Transactions.GetOrderTickets()
                    if x.Symbol == "SPY"
                    and x.Status == OrderStatus.Submitted
                    and x.Tag == "TrailingStopLoss"
                ]

                if len(stopLossOrderTickets) > 0:
                    # Cancel previous trailing stop orders
                    for ticket in stopLossOrderTickets:
                        ticket.Cancel()

                if (
                    self.adx.Current.Value < 20
                ) or (
                    (data.Close < self.ema20.Current.Value)
                    and (self.rsi.Current.Value < 50)
                ):
                    # Place sell order at market price
                    quantity = int(self.Portfolio["SPY"].Quantity)
                    orderTicket = self.MarketOrder("SPY", -quantity)

    def OnOrderEvent(self, orderEvent):
        if orderEvent.Status == OrderStatus.Filled:
            if "TrailingStopLoss" in orderEvent.Order and orderEvent.Order["TrailingStopLoss"] == True:
                self.Debug("Trailing stop loss order filled")

